File: android/src/main/java/com/audiocontext/nodes/AudioBufferSourceNode.kt
package com.audiocontext.nodes

import com.audiocontext.context.BaseAudioContext
import com.audiocontext.parameters.PlaybackParameters
import com.audiocontext.utils.AudioBuffer
import com.audiocontext.utils.Constants

class AudioBufferSourceNode(context: BaseAudioContext) : AudioScheduledSourceNode(context) {
  override var playbackParameters: PlaybackParameters? = null

  private var loop: Boolean = true
    get() = field
  private var buffer: AudioBuffer
    get() = field

  init {
    buffer = AudioBuffer(context.sampleRate, Constants.BUFFER_SIZE, 2)
    playbackParameters = PlaybackParameters(context.getAudioTrack(Constants.BUFFER_SIZE), buffer)
  }

  fun setBuffer(audioBuffer: AudioBuffer) {
    playbackParameters?.let { context.addAudioTrack(it.audioTrack) }
    this.buffer = audioBuffer
    channelCount = audioBuffer.numberOfChannels

    val audioTrack = context.getAudioTrack(2 * buffer.length)
    val buffer = AudioBuffer(buffer.sampleRate, buffer.length, buffer.numberOfChannels)
    playbackParameters = PlaybackParameters(audioTrack, buffer)
  }

  fun setLoop(value: Boolean) {
    loop = value
  }

  override fun fillBuffer(playbackParameters: PlaybackParameters) {
    mixBuffers(playbackParameters)

    if (!loop) {
      isPlaying = false
    } else {
      for (i in 0 until playbackParameters.audioBuffer.numberOfChannels) {
        for (j in 0 until playbackParameters.audioBuffer.length) {
          playbackParameters.audioBuffer.getChannelData(i)[j] = buffer.getChannelData(i)[j % buffer.length]
        }
      }
    }
  }
}

File: packages/react-native-audio-api/ios/nodes/AudioBufferSourceNode/AudioBufferSourceNode.m
#import "AudioBufferSourceNode.h"
#import "AudioContext.h"

@implementation AudioBufferSourceNode

- (instancetype)initWithContext:(AudioContext *)context
{
  self = [super initWithContext:context];
  if (self) {
    _loop = true;
    _bufferIndex = 0;
    self.channelCount = 2;
    _buffer = [[RNAA_AudioBuffer alloc] initWithSampleRate:context.sampleRate
                                                    length:[Constants bufferSize]
                                          numberOfChannels:2];

    self.numberOfOutputs = 1;
    self.numberOfInputs = 0;

    _format = [[AVAudioFormat alloc] initStandardFormatWithSampleRate:_buffer.sampleRate channels:2];
    __weak typeof(self) weakSelf = self;
    _sourceNode = [[AVAudioSourceNode alloc] initWithFormat:_format
                                                renderBlock:^OSStatus(
                                                    BOOL *isSilence,
                                                    const AudioTimeStamp *timestamp,
                                                    AVAudioFrameCount frameCount,
                                                    AudioBufferList *outputData) {
                                                  return [weakSelf renderCallbackWithIsSilence:isSilence
                                                                                     timestamp:timestamp
                                                                                    frameCount:frameCount
                                                                                    outputData:outputData];
                                                }];
    [self.context.audioEngine attachNode:self.sourceNode];
    [self.context.audioEngine connect:self.sourceNode to:self.context.audioEngine.mainMixerNode format:self.format];

    if (!self.context.audioEngine.isRunning) {
      NSError *error = nil;
      if (![self.context.audioEngine startAndReturnError:&error]) {
        NSLog(@"Error starting audio engine: %@", [error localizedDescription]);
      }
    }
  }
  return self;
}

- (void)cleanup
{
  if (self.isPlaying) {
    [self stopPlayback];
  }

  [self.context.audioEngine detachNode:self.sourceNode];
  _format = nil;
}

- (bool)getLoop
{
  return _loop;
}

- (void)setLoop:(bool)loop
{
  _loop = loop;
}

- (void)setBuffer:(RNAA_AudioBuffer *)buffer
{
  _buffer = buffer;
  _mixedBuffer = [_buffer mixWithOutputNumberOfChannels:self.channelCount];
}

- (OSStatus)renderCallbackWithIsSilence:(BOOL *)isSilence
                              timestamp:(const AudioTimeStamp *)timestamp
                             frameCount:(AVAudioFrameCount)frameCount
                             outputData:(AudioBufferList *)outputData
{
  if (outputData->mNumberBuffers < 2) {
    return noErr; // Ensure we have stereo output
  }

  float *leftBuffer = (float *)outputData->mBuffers[0].mData;
  float *rightBuffer = (float *)outputData->mBuffers[1].mData;

  float *leftBufferData = [_mixedBuffer getChannelDataForChannel:0];
  float *rightBufferData = [_mixedBuffer getChannelDataForChannel:1];

  for (int frame = 0; frame < frameCount; frame += 1) {
    // Convert cents to HZ
    if (!self.isPlaying) {
      leftBuffer[frame] = 0;
      rightBuffer[frame] = 0;
      continue;
    }

    leftBuffer[frame] = leftBufferData[_bufferIndex];
    rightBuffer[frame] = rightBufferData[_bufferIndex];

    ++_bufferIndex;

    if (_bufferIndex >= _buffer.length) {
      if (_loop) {
        _bufferIndex = 0;
      } else {
        [self stopPlayback];
        _bufferIndex = 0;
        break;
      }
    }
  }

  [self process:frameCount bufferList:outputData];

  return noErr;
}

@end
